### ISSUE TITLE
Returning named Futures

### ISSUE THREAD

**Issue by tyranron** [sfackler/tokio-openssl#15]
Current [`0.4.0-alpha.6` version](https://docs.rs/tokio-openssl/0.4.0-alpha.6/tokio_openssl/) of this library uses `async fn`, which is obvious and natural way to do async stuff.

However, for low-level code this introduces some challenges as we cannot name the type of returned `Future`, like the following:
```rust
impl<S> ServiceFn for OpensslAcceptor {
    type Ok = SslStream<S>;
    type Error = HandshakeError<S>;
    type Fut = ...;  // <-- we should name the Future here

    fn accept(&self, stream: S) -> Self::Fut {
        tokio_openssl::accept(&self.inner_acceptor, stream)
    }
}
```
which can be resolved only by returning `BoxFuture` at the moment in a price of redudant allocation.

Once this library returns some custom named `Future`s from its public methods, it will allow better costless integration into low-level code of frameworks.

Of course, once `async fn` will be supported in traits, all of this will be unnecessary.

Does the described above makes sense and seems reasonable/feasible for crate maintainers?

**Comment by tyranron** [comment]
For example, [`tokio-rustls`](https://docs.rs/tokio-rustls/0.12.0-alpha.8/tokio_rustls) does so.

**Comment by Arnavion** [comment]
The bigger problem is that the result of `accept` borrows from the `&SslAceptor`. Eg the future returned by OP's code would borrow from `self.inner_acceptor` and thus be bound by `'self`. The pre-async code did not have this problem.

**Comment by sfackler** [comment]
The soon-to-be released 0.6 version will have explicit poll_accept/poll_connect methods.
